stdlib/public/core/UnsafeRawBufferPointer.swift.gyb

```diff
@@ -148,8 +148,21 @@ extension UnsafeRawBufferPointer.Iterator: IteratorProtocol, Sequence {
   public mutating func next() -> UInt8? {
     if _position == _end { return nil }
 
-    let result = _position!.load(as: UInt8.self)
-    _position! += 1
+    // We can do an unchecked unwrap here by borrowing invariants from the pointer.
+    // For a validly constructed buffer pointer, the only way _position can be nil is
+    // if _end is also nil. We checked that case above. Thus, we can safely do an
+    // unchecked unwrap here.
+    //
+    // Additionally, validly constructed buffer pointers also have an _end that is
+    // strictly greater than or equal to _position, and so we do not need to do checked
+    // arithmetic here as we cannot possibly overflow.
+    //
+    // We check these invariants in debug builds to defend against invalidly constructed
+    // pointers.
+    _debugPrecondition(_position! < _end!)
```

Comment:
Couldn't this be hoisted in to the iterator's initializer, as a check that `_position <= _end`? That would give us the same assurances, but IMO would be a bit cleaner; we wouldn't need to talk about 'validly constructed buffer pointer's any more, more like 'validly constructed iterators'. Additionally, it may improve performance in debug builds, which is also important.

Reply:
@stephentyrone Would you like a `!= nil` `_debugPrecondition` as well?

Reply:
I think I'm fine with it as you have it.